specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json

```diff
@@ -0,0 +1,1021 @@
+          "type": "string"
+        },
+        "leaf": {
+          "type": "string"
+        },
+        "leafComponents": {
+          "$ref": "#/definitions/ReceiptLeafComponents"
+        },
+        "nodeId": {
+          "type": "string"
+        },
+        "proof": {
+          "$ref": "#/definitions/ReceiptElementArray"
+        },
+        "root": {
+          "type": "string"
+        },
+        "serviceEndorsements": {
+          "$ref": "#/definitions/PemArray"
+        },
+        "signature": {
+          "type": "string"
+        }
+      },
+      "required": [
+        "signature",
+        "proof",
+        "nodeId"
+      ],
+      "type": "object"
+    },
+    "ReceiptElement": {
+      "properties": {
+        "left": {
+          "type": "string"
+        },
+        "right": {
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "ReceiptElementArray": {
+      "items": {
+        "$ref": "#/definitions/ReceiptElement"
+      },
+      "type": "array"
+    },
+    "ReceiptLeafComponents": {
+      "properties": {
+        "claimsDigest": {
+          "type": "string"
+        },
+        "commitEvidence": {
+          "type": "string"
+        },
+        "writeSetDigest": {
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "RoleAssignment": {
+      "description": "Object for assigning a role to a user.",
+      "required": [
+        "roleName"
+      ],
+      "properties": {
+        "roleName": {
+          "$ref": "#/definitions/LedgerUserRole"
+        },
+        "description": {
+          "description": "Description of the role.",
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "Collection": {
+      "description": "Identifier for collections.",
+      "required": [
+        "collectionId"
+      ],
+      "properties": {
+        "collectionId": {
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "PagedCollections": {
+      "description": "Paginated collections returned in response to a query.",
+      "type": "object",
+      "required": [
+        "collections"
+      ],
+      "properties": {
+        "collections": {
+          "items": {
+            "$ref": "#/definitions/Collection"
+          },
+          "type": "array"
+        },
+        "nextLink": {
+          "description": "Path from which to retrieve the next page of results.",
+          "type": "string"
+        }
+      }
+    },
+    "TransactionId": {
+      "description": "A unique identifier for the state of the ledger. If returned as part of a LedgerEntry, it indicates the state from which the entry was read.",
+      "type": "string"
+    },
+    "TransactionReceipt": {
+      "description": "A receipt certifying the transaction at the specified id.",
+      "required": [
+        "state",
+        "transactionId"
+      ],
+      "properties": {
+        "applicationClaims": {
```

Comment:
Yes, I thought about it when creating application claims and I was initially considering having a `nextLink` useless given that we currently support one write per transaction (and thus only one claim) and support for multiple writes during a transaction is probably not in scope in the near feature. However, we do not have any visibility into how many application claims we may be able to attach to a receipt in the future (even if we only stick to 1 write per transaction), so having a `nextLink` could actually make sense for extensibility purposes.

Technically, the same reasoning should apply on the arrays inside the receipt content (i.e., `proof` and `serviceEndorsements`), but I'm not sure if there was some reason to not have a `nextLink` there, do you happen to know?

I'll probably wait for the API board review before adding a possible `nextLink` field to discuss the use case. However, in case we agree to add it, that would imply we should have the `applicationClaims` array field as a `required` property, while we decided long time ago that it should rather be `optional`. To mitigate, we could potentially add another top structure, such as:

```
{
  "receipt": {
    ...
  },
  "state": "Ready",
  "transactionId": "2.40",
  "applicationClaims": {
    "claims": [
      {
        "claimData": {
          "collectionId": "My collection ID",
          "contents": "My message"
        },
        "secretKey": "Y2Fhc2RzZGFzZGFzZGFzZHNhdmRoYXNqZGhhc2tqZGhhcw==",
        "version": "v1"
      },
      {
        "claimDigest": "221627395c79fc4ffd8e4c2f139bd99aa83c4aa6bde83da39516a1f6bc830667",
        "version": "v1"
      }
    ],
    "nextLink": ""
  }
}
``` 
Let me know if you have any thoughts on this!

Reply:
It was perfectly ok to say a list probably doesn't need to be paged. For example, proof and endorsements are pretty bounded, and application claims makes sense to be bounded as well. Additionally, the receipt is intended to be downloaded all at once, so there is no client-side benefit of "paging over" receipt contents as there might be in a normal collection download where processing of N individual elements is supposed to happen separately (e.g. if you are looping over a collection, you can start processing each item as you receive them instead of waiting up front for all items to be downloaded).

Sorry for the wild goose chase! I think it's unlikely we get asked about it anyway as this list is more buried and is not a top-level collection return. I just wanted to let you know about the possibility so you can be prepared to answer if needed.

Reply:
I'm not too sold on this :D I agree that a receipt is supposed to be downloaded all at once (and making it paginated would not be great for sure), but at the moment we don't know how many claims we could support in the future for a single transaction. In theory, they should be bounded to something (e.g., the number of entries written in the ledger table during a transaction), but in practice we don't have such limit in place (and we don't have a way to derive a reasonable limit to use), so technically nothing prevents a Ledger from returning N claims, where N can be very big, due to a bug or a logical error.

I see this is likely a very small problem and 99% we should be fine with the current approach. But from an API model point of view, maybe it's worth understanding if having a `nextLink` is worth in this scenario, hence my second thoughts in my previous comment. Let's discuss this during the API review and see what the board suggests.

Reply:
We decided that we don't need pagination for application claims as they will be limited in size.